astroquery/mast/core.py

```diff
@@ -1304,7 +1339,69 @@ def _download_files(self, products, base_dir, cache=True):
+                            new_obs_id = obs_id[:-1] + "0"
+                            dataUri = dataUri.replace(obs_id, new_obs_id)
+                            obs_id = new_obs_id
+                            print("WARNING: New URI %s" % dataUri)
+
+                        bucketPath = "hst/public/" + obs_id[:4] + dataUri.replace("mast:HST/product", "")
+
+                        # CAOM LIES!
+                        # We can't use the reported file size
+                        #length = dataProduct["size"]
+                        info_lookup = s3_client.head_object(Bucket='stpubdata', Key=bucketPath, RequestPayer='requester')
+                        length = info_lookup["ContentLength"]
+
+                        if cache and os.path.exists(localPath):
+                            if length is not None:
+                                statinfo = os.stat(localPath)
+                                if statinfo.st_size != length:
+                                    log.warning("Found cached file {0} with size {1} that is "
+                                                "different from expected size {2}"
+                                                .format(localPath,
+                                                        statinfo.st_size,
+                                                    length))
+                                else:
+                                    log.info("Found cached file {0} with expected size {1}."
+                                             .format(localPath, statinfo.st_size))
+                                    s3_fetch_ok = True
+
+                        if not s3_fetch_ok:
+                            with ProgressBarOrSpinner(length, ('Downloading URL s3://stpubdata/{0} to {1} ...'.format(bucketPath, localPath))) as pb:
+                                global bytes_read
+                                bytes_read = 0
+
+                                progress_lock = threading.Lock()
+
+                                def progress_callback(numbytes):
+                                    global bytes_read
+
+                                    # This callback can be called in multiple threads
+                                    # Access to updating the console needs to be locked
+                                    with progress_lock:
+                                        bytes_read += numbytes
+                                        pb.update(bytes_read)
+
+                                bkt.download_file(bucketPath, localPath, ExtraArgs={"RequestPayer": "requester"}, Callback=progress_callback)
+
+                            s3_fetch_ok = True
+                    except Exception as ex:
+                        print("Error pulling from S3 bucket: %s" % ex)
+                        print("Falling back to mast download...")
+
+                if not s3_fetch_ok:
+                    self._download_file(dataUrl, localPath, cache=cache)
```

Comment:
while this approach is fine (i.e., `s3_fetch_ok = False`, then `if not s3_fetch_ok:`), but why not just use a try/except block for this whole thing?

Reply:
It might also make sense to move the whole download-from-s3 code block to its own private function, e.g., `self._download_file_from_s3`, so that the try-except block is easy to read

Reply:
I like the separate function idea, then there is nice parity with the `self._download_file` function.